File: baekjoon/11725/11725.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

const int MAX_LEN = 100000 + 1;

void bfs(vector<int> graph[], int visited[], const int& startVertex) {

	queue<int> procedure;
	visited[startVertex] = startVertex;
	procedure.push(startVertex);

	while (procedure.empty() == false) {
		const int currentVertex = procedure.front();
		procedure.pop();

		for (const auto& iter : graph[currentVertex]) {
			if (visited[iter] == 0) {
				visited[iter] = currentVertex;
				procedure.push(iter);
			}
		}
	}
}

void run() {
	int N = 0;
	cin >> N;

	vector<int> graph[MAX_LEN];
	for (int i = 1; i < N; ++i) {
		int vertex1 = 0, vertex2 = 0;
		cin >> vertex1 >> vertex2;

		graph[vertex1].push_back(vertex2);
		graph[vertex2].push_back(vertex1);
	}

	int visited[MAX_LEN] = {
		0,
	};

	bfs(graph, visited, 1);

	for (int i = 2; i <= N; ++i) {
		cout << visited[i] << "\n";
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	run();

	return 0;
}

File: baekjoon/11725/11725.java
import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;
import java.util.Vector;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());

        Vector<Integer>[] graph = new Vector[(N + 1) * 2];
        for (int i = 1; i <= N; ++i) {
            graph[i] = new Vector<>();
        }
        for (int i = 1; i < N; ++i) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");

            int vertex1 = Integer.parseInt(st.nextToken());
            int vertex2 = Integer.parseInt(st.nextToken());

            graph[vertex1].add(vertex2);
            graph[vertex2].add(vertex1);
        }

        int[] visited = new int[N + 1];

        bfs(graph, visited, 1);

        for (int i = 2; i <= N; ++i) {
            bw.write(visited[i] + "\n");
        }

        br.close();
        bw.flush();
        bw.close();
    }

    private static void bfs(Vector<Integer>[] graph, int[] visited, int startVertex) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(startVertex);

        while (queue.isEmpty() == false) {
            int currentVertex = queue.poll();

            for (int i : graph[currentVertex]) {
                if (visited[i] != 0) {
                    continue;
                }

                visited[i] = currentVertex;
                queue.add(i);
            }
        }
    }
}
